[37107] render tabs in component

oliverguenther opened the pull request:
- [x] Renders all core work package tabs in one component
- [x] Register core tabs dynamically, requires passing in the route to determine whether we're in split or full view
- [x] Allow tab counts to be defined dynamically

https://community.openproject.org/wp/37107

oliverguenther committed:
Render work package tabs in one component

b12f commented on frontend/src/app/components/wp-tabs/components/wp-tabs/wp-tabs.component.html:
```diff
@@ -1,5 +1,68 @@
+      [uiParams]="{workPackageId: workPackage.id}"
+      uiSrefActive="selected"
+  >
+    <a href="" [textContent]="text.tabs.overview"></a>
```
Aren't the uiSref and uiParams on the wrong element here?

oliverguenther replied:
The problem here is the style coming from a generic "tabrow" style component, which styles the `li` to have a padding. Placing uiSref on the link then would not trigger on a click outside of the actual text.

Ideally we also change the styling of the `tabrow` but I'm not sure that this should be in the scope of this PR

oliverguenther committed:
Move standard tabs into tab service

oliverguenther committed:
Move tab-count styles mostly into its own component style

oliverguenther committed:
Fix frontend specs

oliverguenther committed:
Remove unused texts

oliverguenther committed:
Fix routing to split view from button

oliverguenther committed:
Fix context menu not opening the tabbed view

oliverguenther committed:
Fix watcher tab counter not receiving an rxjs subscription

oliverguenther committed:
Fix selector

oliverguenther committed:
Fix clicking on relation link to fullscreen

ulferts reviewed (changes requested):
Somebody else should probably take another look a the code. My take on it would be, that it is fine but there might be more qualified opinions.

From a functionality perspective I find it working very well except for one edge case:
* Reduce the width of the split view
* Have watchers and relations
* Have the GitHub tab by enabling the module
* Switch to German:
![image](https://user-images.githubusercontent.com/617519/117979184-f0864480-b332-11eb-8aa3-9fc470536773.png)

It also fails on mobile but that page has so many bugs that I don't know if it counts:

![image](https://user-images.githubusercontent.com/617519/117979430-2e836880-b333-11eb-9d0e-5b23ca1ddba0.png)

oliverguenther committed:
Make WP tabs an instance of scrollable-tabs

oliverguenther committed:
Convert the tabs from .tabrow to its own BEM block

oliverguenther committed:
Refactor other usages of tabrow into op-tab-row

oliverguenther committed:
Fix frontend specs

oliverguenther committed:
Make wp modal tabs use scrollable-tabs

oliverguenther committed:
Fix disabled tab selectors

oliverguenther committed:
Restore scrollability of tabs

b12f commented on frontend/src/app/components/wp-fast-table/builders/ui-state-link-builder.ts:
```diff
@@ -4,32 +4,37 @@ import { StateService } from '@uirouter/core';
-    const selectedTabIdentifier = this.$state.params?.tabIdentifier;
-
+    let tabState:string;
+    let tabIdentifier:string;
```
I prefer only using `const` when possible, and I'd advocate for two ternary operations here, or perhaps wrapping it in an IIFE

oliverguenther replied:
I feel like an IIFE would make this block more complicated than it needs to be. It's an IF that applies two variables and with two local vars. I find my version easier to read but would accept both versions. I couldn't find any specific entry to the style guide either

b12f replied:
I'll accept it for now but the next iteration of the style guide bans `let` ;)

b12f commented on frontend/src/app/components/wp-tabs/components/wp-tabs/wp-tabs.component.html:
```diff
@@ -1,5 +1,27 @@
+  class="op-work-package-tabs">
+    <ng-container slot="actions" *ngIf="view === 'split'">
+      <li class="tab-icon">
+        <accessible-by-keyboard
```
I've read it's code and I'm really unsure what advantage this would bring over using `<a role="tab">`, especially in this location.

b12f replied:
I'm not an a11y expert by any means though

oliverguenther replied:
The main reason for accessible-by-keyboard was the ability to trigger click handlers with either space or enter. It's a WAI aria requirement that doesn't work with links unless you specify an href.

- `a href` or `a href"#"` or `a href=""` allows to trigger with enter, but not with space. `role=button` doesn't change anything there, but allows to omit the `href`

- `<button>` with and without `role=button` (which is the default anyway) allows to trigger with space and enter

Of course you could always add your own listeners, but that's exactly what accessible-by-keyboard (and accessibleClick) is for

So I don't see a reason why we cannot make all of those cases a `button` then.  Let's do some more experiments and open a code maintenance ticket then to deprecate `accessible-by-keyboard` if it solves all cases. I would however do these consistency changes in a separate PR, as this again just mostly moved code

b12f replied:
I always thought there were specific a11y reasons why browsers by default have different ux for buttons and anchors, which is why I tend to default to browser standards (e.g. links are anchors, actions are buttons), but if every link/button requires to be activated by both perhaps we can rethink the component as a directive instead. This is future stuff though, I agree

b12f commented on frontend/src/app/components/wp-tabs/services/wp-tabs/wp-tabs.service.ts:
```diff
@@ -1,25 +1,85 @@
+    return [
+      {
+        component: WorkPackageOverviewTabComponent,
+        name: this.I18n.t('js.work_packages.tabs.overview'),
```
Not sure if this service should know about any specific tab sets at all.

oliverguenther replied:
It's a WP single view specific service, so in this case I think this is the best place to put it. You could extract it into a separate APP_INITIALIZER, but that only makes it harder to understand where the tabs are coming from.

b12f commented on frontend/src/app/modules/autocompleter/te-work-package-autocompleter/te-work-package-autocompleter.component.html:
```diff
@@ -1,33 +1,44 @@
+            (click)="setMode('all')"
+        >
+          <a
+              href="#"
```
This should probably be a `button role="tab" type="button"` and the click listener should be set on this element isntead of the `li`

b12f replied:
Same for the code below

oliverguenther replied:
👍 on role=tab with the parent being a tablist. I'll look into it for all mentioned cases

b12f commented on frontend/src/app/modules/boards/board/configuration-modal/board-configuration.modal.html:
```diff
@@ -1,25 +1,27 @@
-            (click)="switchTo(tab.name)"
-            href
+              class="op-tab-row--link"
+              role="button"
```
Any reason you preferred `a role="button"` over a `button type="button"` ?

b12f commented on frontend/src/app/modules/common/tabs/scrollable-tabs/scrollable-tabs.component.html:
```diff
@@ -1,19 +1,61 @@
+          [attr.data-tab-id]="tab.id"
+          [attr.data-qa-tab-id]="tab.id"
+          [attr.data-qa-tab-selected]="tab.id === currentTabId || undefined"
+          [attr.data-qa-tab-disabled]="tab.disable || undefined"
```
Yeah I can see how this becomes a struggle. Maybe we can do something like `data-qa="/** json stuff **/"`, though that wouldn't be much better probably

b12f commented on frontend/src/app/modules/common/tabs/scrollable-tabs/scrollable-tabs.component.html:
```diff
@@ -1,19 +1,61 @@
+    </ul>
+    <div #scrollLeftBtn [hidden]="hideLeftButton"
+         class="op-scrollable-tabs--button -left"
+         (click)="scrollLeft()"
```
probably want a `button type="button"` for this element and the one below, too

b12f commented on frontend/src/app/modules/common/tabs/tab-badges/tab-count.component.ts:
```diff
@@ -0,0 +1,12 @@
+@Component({
+  selector: 'op-tab-count',
+  templateUrl: './tab-count.html',
+  styleUrls: ['./tab-count.sass'],
```
`.component.` for both html template and sass would be nice. I use it to separate template-specific sass from general BEM/util stuff.

b12f commented on frontend/src/global_styles/content/_scrollable_tabs.sass:
```diff
@@ -0,0 +1,54 @@
+  border-bottom: 1px solid #DDDDDD
+  margin-bottom: 1rem
+
+  &.-narrow
```
this is completely new code, afaik. In the spirit of BEM, I recommend setting modifiers by using `&_narrow`, that way a modifier selector is still only 1 class of specificity (the style your using results in `.op-scrollable-tab.-narrow`, while you really want `.op-scrollable-tab_narrow` as one class)

oliverguenther replied:
The code was moved (and partially changed), but I can still rename the modifier.

b12f commented on frontend/src/global_styles/content/_scrollable_tabs.sass:
```diff
@@ -0,0 +1,54 @@
+    background: white
+    color: var(--light-gray)
+    font-size: 0.7rem
+    &.-left
```
Same here, one class selectors with `_` as a separator for modifiers

``` 
.op-tabs // block
.op-tabs--tab // element
.op-tabs_big // block modifier
.op-tabs--tab_big // element modifier
```

oliverguenther committed:
Fix import and change detection of graph modal

oliverguenther committed:
Fix selector in wp relations spec

oliverguenther committed:
Rename tab-count to match component style

oliverguenther committed:
Remove narrow style

oliverguenther committed:
Fix bem modifier style

ulferts reviewed (changes requested):
I did not look at the code, only at the UI:

![image](https://user-images.githubusercontent.com/617519/118613159-c6afa080-b7be-11eb-83db-aee30c63228d.png)

* The blue bar indicating the active tab is missing
* The text is too small in my opinion. 
* The count indicators are cut off at the bottom
* The scroll arrows to the right should be removed when not needed. As the scrollability handles an edge case, having disabled arrows at the top is distracting without need.
* The line below the tabs is broken below the arrow keys

I did not find a way to reproduce the headings getting squashed, though.

Sorry for the barrage of naggings.

oliverguenther committed:
Fix styles of scrollable tabs

oliverguenther commented:
> I did not look at the code, only at the UI:
> 
> ![image](https://user-images.githubusercontent.com/617519/118613159-c6afa080-b7be-11eb-83db-aee30c63228d.png)
> 
> * The blue bar indicating the active tab is missing


fixed, thanks

> * The text is too small in my opinion.

It now uses the scrollable-tabs which were fixed at 14px before already. It's now consistent at that level (matches other tabs such as the user admin tabs). So I'd like to keep it consistent at that size for now instead of introducing another size for the details tabs again.

> * The count indicators are cut off at the bottom

fixed, thanks.

> * The scroll arrows to the right should be removed when not needed. As the scrollability handles an edge case, having disabled arrows at the top is distracting without need.

The arrows are only shown when needed, however resizing e.g., the details bar does not immediately update that logic. We'd need to hook into the resizer to know when the page is resized. I can look into it, but don't find it too important for now. With the default width of the split view (reset your local storage to test this) and the github module disabled, the arrows are not shown.

> * The line below the tabs is broken below the arrow keys

fixed

oliverguenther commented:
Hi @b12f , could I ask you to have another go at the code review ? Thanks :bow:

oliverguenther committed:
Fix tab for time entry autocompleter

b12f reviewed (approved):
LGTM
